### ISSUE TITLE
Rule Change: update default of `no-shadow.hoist` option to be `"functions-and-types"` to match `typescript-eslint`

### ISSUE THREAD

**Issue by JstnMcBrd** [eslint/eslint#20776]
### What rule do you want to change?

no-shadow

### What change do you want to make?

Generate more warnings

### How do you think the change should be implemented?

A new default behavior

### Example code

```ts
/*eslint no-shadow: ["error", { "hoist": "functions-and-types" }]*/

// types
type Bar<Foo> = 1;
type Foo = 1;

// functions
if (true) {
  const b = 6;
}

function b() {}
```

### What does the rule currently do for this code?

Only warns about the shadowed function.

### What will the rule do after it's changed?

Warn about the shadowed type too.

### Participation

- [x] I am willing to submit a pull request to implement this change.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

### Additional comments

This is a follow-up on
- #19565

which is related to
- #19173

---

Here is my current setup for enabling `no-shadow` for TypeScript:

```ts
{
  // ... extending a config that uses "no-shadow": "error" ...
  rules: {
    // as recommended by https://typescript-eslint.io/rules/no-shadow#how-to-use
    "no-shadow": "off",
    "@typescript-eslint/no-shadow": "error"
  }
}
```

I was excited to see support for TypeScript in `no-shadow` was added in #19565, because then (in theory) I should be able to remove this override.

But upon closer examination, I realized that the `hoist` option default was not updated - it is still `"functions"`, when for typescript-eslint it is `"functions-and-types"`.
- https://github.com/eslint/eslint/blob/main/lib/rules/no-shadow.js#L65
- https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/src/rules/no-shadow.ts#L98

So by relying on eslint's core rule (without modification), I would be potentially missing some issues.

Of course, I can just customize the rule with `{ "hoist": "functions-and-types" }` as a workaround. But I always prefer to reduce excess configuration when possible - hence this request :)

<br>

I wasn't sure if the difference was intentional. I figure it's odd that this small discrepancy remains after all the effort to bring `no-shadow` into alignment. Also, if the rule purportedly supports TypeScript by default, it's odd that the default behavior would miss some of the issues in TypeScript.

But even if it is intentional, I'd like to advocate for updating it. I think it'd be a low-risk change, it irons out a small remaining wrinkle, slightly improves DevX, furthers the goals of #19173, etc. Even if needs to wait until a major update, I still think it would be worthwhile.

Let me know what you think!

**Comment by DMartens** [comment]
I agree to changing the default value to `function-and-types` in the next major version for a different reason: for both functions and types the order of declaration does not matter.
I tried to uncover why we did not adapt the default value but I could find not discussion related to this, so this seems like an oversight.
The option value `hoist: function-and-types` were added just before we adopted the changes.

**Comment by Tanujkanti4441** [comment]
I think it's good to update `hoist` default to `function-and-types` so users don't need to update their config to `/*eslint no-shadow: ["error", { "hoist": "functions-and-types" }]*/` if they are switching from `@typescript-eslint/no-shadow`. So, we can avoid cases where users can assume the default value for `hoist` is the same in both `@typescript-eslint/no-shadow` and ESLint's `no-shadow`, but it's not.

Let's here from other members @eslint/eslint-team.

**Comment by fasttime** [comment]
Since this would be a breaking change per our policies, we can defer the decision to the next major release, and gather feedback in the meantime.
